### PR TITLE
http: optimize header directives by avoiding HttpHeader.unapply

### PR DESCRIPTION
**Pull request by nitikagarw**
Related PR: https://github.com/akka/akka-http/pull/3591
Please refer to the above PR description for more details

### CODE REVIEW

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 1]
```diff
@@ -133,7 +133,7 @@ trait HeaderDirectives {
    * @group header
    */
   def optionalHeaderValueByName(headerName: String): Directive1[Option[String]] = {
-    val lowerCaseName = headerName.toLowerCase
+    val lowerCaseName = headerName.toRootLowerCase
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala:136]: Good catch!

**Review by jrudolph — APPROVED** [review]
LGTM, thanks a lot :)

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala (hunk 1 above)
Comment (raboof) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala:136]: This makes sense to me, though I wonder if there aren't more places where we should do this

**Review by raboof — APPROVED** [review]

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala:136]: Definitely, but that one was related to the previous PR, so makes sense to do that here.

**Comment by jrudolph** [comment]
Thanks again, @nitikagarw!

**Pull request merged**